### PR TITLE
Keep the config intact as previous state

### PR DESCRIPTION
**Pull request by dasmanas**
## Description

Fixes pyrsia/pyrsia#1273

This PR keeps the config in original state after test case run completed.

## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/community/get_involved/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).

### Code Contributions

- [x] I've built the code `cargo build --all-targets` successfully.
- [x] I've run the unit tests `cargo test --workspace` and everything passes.
- [x] I've made sure my rust toolchain is current `rustup update`.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -4147,18 +4147,18 @@ checksum = "949517c0cf1bf4ee812e2e07e08ab448e3ae0d23472aee8a06c985f0c8815b16"
 
 [[package]]
 name = "thiserror"
-version = "1.0.35"
```
Comment (jperedadnr) [inline Cargo.lock:4150]: why the change here, and not in `Cargo.toml`?

File: pyrsia_cli/src/cli/handlers.rs [hunk 2]
```diff
@@ -357,6 +365,7 @@ mod tests {
 
     #[test]
     fn test_valid_config_edit() {
+        let existing_cli_config = config::get_config().unwrap();
```
Comment (jperedadnr) [inline pyrsia_cli/src/cli/handlers.rs:368]: This works fine now, of course. 

But it might be a different option to use some different file and not the one "in production"?

So far we are using the same default file via `get_config()` that uses `confy::load(CONF_FILE, None)`, which in turn calls `config_name.into().unwrap_or("default-config")`, but that has room for improvement.

If we use for instance, `confy::load(CONF_FILE, TEST_FILE)`, then we won't override the user configuration.
Does this make sense to you?

File: pyrsia_cli/src/cli/handlers.rs [hunk 3]
```diff
@@ -365,13 +374,21 @@ mod tests {
...
         };
-        config_edit(host_name, port, diskspace);
+        let config_edit_result = config_edit(host_name, port, diskspace);
         let updated_cli_config = config::get_config().unwrap();
+        match config_edit_result {
+            Ok(_) => {
+                //restore the config to original state after test
+                let _restore_config = config::add_config(existing_cli_config);
+            }
+            Err(_) => {}
```
Comment (jperedadnr) [inline pyrsia_cli/src/cli/handlers.rs:384]: If there is an error, shouldn't we also restore the original file? And shouldn't we log such error?

File: pyrsia_cli/src/cli/handlers.rs [hunk 4]
```diff
@@ -380,8 +397,15 @@ mod tests {
...
```
Comment (jperedadnr) [inline pyrsia_cli/src/cli/handlers.rs:407]: Same here

File: Cargo.lock (hunk 1 above)
Comment (dasmanas) [reply Cargo.lock:4150]: I can see the file `src/blockchain/Cargo.toml` in the commit. It is at the end.

File: pyrsia_cli/src/cli/handlers.rs (hunk 2 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs:368]: In the recent latest version [`confy::load`](https://docs.rs/confy/latest/confy/fn.load.html) changed the function signature. It introduced a new arg `app_name`. The use case more like `confy::load("pyrsia", "cli")`, `confy::load("pyrsia", "node")` or `confy::load("pyrsia", "some_other_module")` instead of `confy::load(CONF_FILE, None)` what we are currently using.

File: pyrsia_cli/src/cli/handlers.rs (hunk 3 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs:384]: `config_edit_result` is holding the outcome of `pub fn config_edit(host_name: Option<String>, port: Option<String>, diskspace: Option<String>,) -> Result<(), Error>`. If there is error at the time of doing edit, we have nothing to restore back. Because anyway edit commit is not successful.

File: pyrsia_cli/src/cli/handlers.rs (hunk 4 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs:407]: `config_edit_result` is holding the outcome of `pub fn config_edit(host_name: Option<String>, port: Option<String>, diskspace: Option<String>,) -> Result<(), Error>`. If there is error at the time of doing edit, we have nothing to restore back. Because anyway edit commit is not successful.

File: Cargo.lock (hunk 1 above)
Comment (jperedadnr) [reply Cargo.lock:4159]: yes, but don't you need to change the main [`Cargo.toml` ](https://github.com/pyrsia/pyrsia/blob/main/Cargo.toml) at the root level? 
Imo, blockchain has nothing to do with this PR...

Comment (dasmanas) [reply Cargo.lock:4159]: Reverted back the `Cargo.toml`. Change not required. `cargo clean` did the trick.

File: pyrsia_cli/src/cli/handlers.rs (hunk 2 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs:368]: new issue is created https://github.com/pyrsia/pyrsia/issues/1285

File: Cargo.lock [hunk 5]
```diff
@@ -4156,18 +4156,18 @@ checksum = "949517c0cf1bf4ee812e2e07e08ab448e3ae0d23472aee8a06c985f0c8815b16"
...
+version = "1.0.37"
```
Comment (jperedadnr) [inline Cargo.lock:4159]: Don't you need to reverse these changes too?

File: pyrsia_cli/src/cli/handlers.rs (hunk 3 above)
Comment (jperedadnr) [reply pyrsia_cli/src/cli/handlers.rs:384]: Then you can skip the `match` and simply use
```
if config_edit_result.is_ok() {
   // restore the ...
}
```

File: pyrsia_cli/src/cli/handlers.rs (hunk 4 above)
Comment (jperedadnr) [reply pyrsia_cli/src/cli/handlers.rs:407]: same here

File: Cargo.lock (hunk 5 above)
Comment (dasmanas) [reply Cargo.lock:4159]: interesting thing is `Cargo.lock` is not getting change even though I change the `Cargo.toml`. I shouldn't change it manually too. Let me see what is going on.

Comment (dasmanas) [reply Cargo.lock:4159]: `thiserror = "1.0.35"` means the range is `>=1.0.35, <2.0.0`. Explained under [Specifying dependencies from crates.io](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio)

**Although it looks like a specific version of the time crate, it actually specifies a range of versions and allows [SemVer](https://semver.org/) compatible updates. An update is allowed if the new version number does not modify the left-most non-zero digit in the major, minor, patch grouping.**

File: pyrsia_cli/src/cli/handlers.rs [hunk 6]
```diff
@@ -93,14 +98,17 @@ pub fn config_edit(host_name: Option<String>, port: Option<String>, diskspace: O
...
+                Ok(())
             } else {
                 errors.into_iter().for_each(|x| println!("{}", x));
-            }
+                Err(anyhow!("Invalid pyrsia config"))
+            };
         }
         Err(error) => {
             println!("Error Saving Node Configuration:       {}", error);
+            Err(anyhow!("Error Saving Node Configuration:       {}", error))
```
Comment (tiainen) [inline pyrsia_cli/src/cli/handlers.rs:109]: Instead of both printing and returning the error, we can just return the error and print it where the `config_edit` method is invoked?

File: pyrsia_cli/src/cli/handlers.rs [hunk 7]
```diff
@@ -93,14 +98,17 @@ pub fn config_edit(host_name: Option<String>, port: Option<String>, diskspace: O
                         println!("Error Saving Node Configuration:       {}", error);
```
Comment (tiainen) [inline pyrsia_cli/src/cli/handlers.rs:98]: Return this as an error too?

File: pyrsia_cli/src/cli/handlers.rs (hunk 6 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs]: changed as suggested.

File: pyrsia_cli/src/cli/handlers.rs (hunk 7 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs]: changed as suggested.

**Comment by tiainen** [comment]
@dasmanas can you replace the `match` with an `if` as suggested by José? Other than that, the PR looks fine to me.

**Comment by dasmanas** [comment]
> @dasmanas can you replace the `match` with an `if` in the tests as [suggested by José](https://github.com/pyrsia/pyrsia/pull/1279/files#r1020539779)? Other than that, the PR looks fine to me.

changed as per suggestion.

File: pyrsia_cli/src/cli/handlers.rs (hunk 3 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs]: `if config_edit_result.is_ok()` done

File: pyrsia_cli/src/cli/handlers.rs (hunk 4 above)
Comment (dasmanas) [reply pyrsia_cli/src/cli/handlers.rs]: if config_edit_result.is_ok()

**Comment by dasmanas** [comment]
@tiainen , @jperedadnr PR is ready for review.

codecov/patch — 36.36% of diff hit (target 47.50%) -> This only can be fixed with the introduction of `mockall` in pyrsia code base and some code refactoring is required. As some bare functions need to go under `struct` or `impl` to mock behavior or outcome

File: Cargo.lock (hunk 1 above)
Comment (jperedadnr) [reply Cargo.lock:4150]: You did revert the `src/blockchain/Cargo.toml` file, that was fine.

But why `Cargo.lock` is still modified? That would happen if you modify the main `Cargo.toml` with:

```
-thiserror = "1.0.35"
+thiserror = "1.0.37"
``` 

Can you please either apply this change to `Cargo.toml` or reset the current change from `Cargo.lock`?

Comment (dasmanas) [reply Cargo.lock]: Done.

**Review by tiainen — APPROVED** [review]

**Review by jperedadnr — APPROVED** [review]
